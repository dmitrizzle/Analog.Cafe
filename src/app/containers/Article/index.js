// tools
import React from "react"
import { Editor, Raw } from "slate"
import Helmet from "react-helmet"

// redux & state
import { connect } from "react-redux"
import { fetchPage } from "../../../actions/articleActions"
import {
  ROUTE_ARTICLE_API,
  ROUTE_ARTICLE_DIR
} from "../../../constants/article"
import { ROUTE_AUTHOR_API } from "../../../constants/author"
import {
  ROUTE_APP_PRODUCTION_DOMAIN_NAME,
  ROUTE_APP_PRODUCTION_DOMAIN_PROTOCOL
} from "../../../constants/app"

import { schema } from "../Composer/containers/ContentEditor/schema"

// components
import Heading from "../../components/ArticleHeading"
import { ModalDispatch } from "../Modal"
import {
  Section,
  Article as ArticleElement,
  Byline
} from "../../components/ArticleStyles"

// render
class Article extends React.PureComponent {
  fetchPage = () => {
    // do not fetch pages unless they are located in /zine dir
    // otherwise on unmount the component will try to load any page, and return 404 errors
    if (!this.props.history.location.pathname.includes(ROUTE_ARTICLE_DIR))
      return

    this.props.fetchPage({
      url:
        ROUTE_ARTICLE_API +
        this.props.history.location.pathname.replace(ROUTE_ARTICLE_DIR, "")
    })
  }
  componentDidMount = () => {
    this.unlisten = this.props.history.listen(location => this.fetchPage())
    this.fetchPage()
  }
  componentWillUnmount = () => {
    this.unlisten()
  }
  render = () => {
    return (
      <ArticleElement>
        <Helmet>
          <title>{this.props.article.title}</title>
          <meta name="description" content={this.props.article.summary} />
          <meta
            property="og:image"
            content={
              this.props.article.poster &&
              ROUTE_APP_PRODUCTION_DOMAIN_PROTOCOL +
                ROUTE_APP_PRODUCTION_DOMAIN_NAME +
                this.props.article.poster.medium
            }
          />
        </Helmet>
        <Heading
          pageTitle={this.props.article.title}
          pageSubtitle={this.props.article.subtitle}
          title={this.props.article.error && this.props.article.error}
        >
          {this.props.article.status === "published" && (
            <Byline>
              by{" "}
              <ModalDispatch
                with={{
                  request: {
                    url: ROUTE_AUTHOR_API + "/" + this.props.article.author.id
                  }
                }}
              >
                {this.props.article.author.name}
              </ModalDispatch>
            </Byline>
          )}
        </Heading>
        <Section articleStatus={this.props.article.status}>
          <Editor
            readOnly={true}
            state={Raw.deserialize(this.props.article.content.raw, {
              terse: true
            })}
            schema={schema}
          />
        </Section>
      </ArticleElement>
    )
  }
}

// connet with redux
const mapStateToProps = state => {
  return {
    article: state.article
  }
}
const mapDispatchToProps = dispatch => {
  return {
    fetchPage: request => {
      dispatch(fetchPage(request))
    }
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(Article)
// NOTE: withRouter() props inherited from /components/_screens/AppRoutes
