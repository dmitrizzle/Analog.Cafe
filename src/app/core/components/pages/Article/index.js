import { connect } from "react-redux"
import LazyLoad from "react-lazyload"
import Loadable from "react-loadable"
import React from "react"

import { ROUTE_TAGS } from "../../../constants/routes-list"
import {
  ROUTE_URL_ARTICLES,
  ROUTE_URL_SUBMISSIONS
} from "../../../constants/routes-article"
import {
  fetchArticlePage,
  setArticlePage
} from "../../../store/actions-article"
import { getSubmissionOrArticleRoute } from "../../../utils/routes-article"
import { getTitleFromSlug } from "../../../utils/messages-"
import ArticleHeader from "./components/ArticleHeader"
import ArticleSection from "./components/ArticleSection"
import ArticleWrapper from "./components/ArticleWrapper"
import MetaTags from "../../vignettes/MetaTags"
import renderArticle from "../../../utils/render-article"

const ArticleActions = Loadable({
  loader: () => import("../../controls/ArticleActions"),
  loading: () => null,
  delay: 100
})
class Article extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      subscribeForm: false,
      tag: {
        name: "Post",
        route: "/"
      },
      publicationStatus: this.props.article.status,
      selection: {
        leftOffset: 0,
        topOffset: 0
      }
    }
  }
  fetchArticlePage = () => {
    if (
      !this.props.history.location.pathname.includes(ROUTE_URL_ARTICLES) &&
      !this.props.history.location.pathname.includes(ROUTE_URL_SUBMISSIONS)
    )
      return
    this.props.fetchArticlePage({
      url:
        getSubmissionOrArticleRoute(this.props.history.location.pathname)
          .apiRoute +
        this.props.history.location.pathname.replace(
          getSubmissionOrArticleRoute(this.props.history.location.pathname)
            .pathname,
          ""
        )
    })
    this.setState({
      subscribeForm: false,
      publicationStatus: this.props.article.status
    })
  }
  makeTag = props => {
    const tag = props.article.tag
    if (typeof tag === "undefined") return
    this.setState({
      ...this.state,
      tag: {
        name: getTitleFromSlug(tag, {
          titleCase: false,
          capitalize: true,
          smartTagFromImageCount: props.article.stats
            ? props.article.stats.images
            : 0
        }),
        route: Object.keys(ROUTE_TAGS).find(key => ROUTE_TAGS[key] === tag)
      }
    })
  }
  componentDidMount = () => {
    this.unlisten = this.props.history.listen(() => this.fetchArticlePage())
    this.fetchArticlePage()
    this.makeTag(this.props)
    this.setState({
      adminControls: this.props.user.info.role === "admin"
    })
  }
  componentWillReceiveProps = nextProps => {
    this.makeTag(nextProps)
    this.setState({
      adminControls: nextProps.user.info.role === "admin",
      publicationStatus: nextProps.article.status
    })
  }
  componentWillUnmount = () => {
    this.unlisten()
  }
  handleSubscribeFormCallback = value => {
    this.setState({
      subscribeForm: !value
    })
  }

  handleMouseUp = event => {
    const selection = window.getSelection()
    const range = selection.getRangeAt(0)
    const rect = range.getBoundingClientRect()
    const menu = {
      offsetWidth: 100,
      offsetHeight: 20
    }
    const leftOffset =
      rect.left + window.scrollX - menu.offsetWidth / 2 + rect.width / 2
    const topOffset = rect.top + window.scrollY - menu.offsetHeight + 3
    this.setState({
      selection: {
        leftOffset,
        topOffset
      }
    })
  }
  shouldComponentUpdate(nextProps, nextState) {
    if (this.props !== nextProps) {
      return true
    } else {
      return false
    }
  }

  render = () => {
    return (
      <ArticleWrapper>
        <MetaTags
          metaTitle={
            this.props.article.title +
            (this.props.article.subtitle
              ? ` (${this.props.article.subtitle})`
              : "")
          }
          metaDescription={this.props.article.summary}
          metaImage={this.props.article.poster}
          metaPostDate={
            this.props.article.date && this.props.article.date.published
          }
          metaEditDate={
            this.props.article.date && this.props.article.date.updated
          }
          metaAuthors={this.props.article.authors}
          metaSlug={this.props.article.slug}
          metaArticleSchema
        />
        <ArticleHeader
          article={this.props.article}
          stateAdminControls={this.state.adminControls}
          stateTag={this.state.tag}
        />
        <ArticleSection
          articleStatus={this.props.article.status}
          onMouseUp={this.handleMouseUp}
        >
          <div
            style={{
              width: "100px",
              height: "20px",
              background: "#000",
              position: "absolute",
              top: `${this.state.selection.topOffset}px`,
              left: `${this.state.selection.leftOffset}px`
            }}
          />
          {renderArticle(this.props.article.content.raw)}
          {this.props.article.poster &&
            this.props.article.submittedBy && (
              <LazyLoad
                once
                offset={300}
                height={"100%"}
                key={`ArticleActions_${this.props.article.slug}`}
              >
                <ArticleActions
                  user={this.props.user}
                  subscribeFormCallback={this.handleSubscribeFormCallback}
                  subscribeForm={this.state.subscribeForm}
                  nextArticle={this.props.article.next}
                  thisArticle={this.props.article.slug}
                  thisArticlePostDate={
                    this.props.article.date && this.props.article.date.published
                  }
                  thisArticleEditDate={
                    this.props.article.date && this.props.article.date.updated
                  }
                  nextArticleHeading={nextArticleHeading =>
                    this.props.setArticlePage({
                      title: nextArticleHeading.title,
                      subtitle: nextArticleHeading.subtitle,
                      authors: nextArticleHeading.authors,
                      slug: nextArticleHeading.slug,
                      poster: nextArticleHeading.poster,
                      tag: nextArticleHeading.tag
                    })
                  }
                />
              </LazyLoad>
            )}
        </ArticleSection>
      </ArticleWrapper>
    )
  }
}
const mapStateToProps = state => {
  return {
    article: state.article,
    user: state.user
  }
}
const mapDispatchToProps = dispatch => {
  return {
    fetchArticlePage: request => {
      dispatch(fetchArticlePage(request))
    },
    setArticlePage: nextArticle => {
      dispatch(setArticlePage(nextArticle))
    }
  }
}
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Article)
// NOTE: withRouter() props inherited from /components/_screens/AppRoutes
