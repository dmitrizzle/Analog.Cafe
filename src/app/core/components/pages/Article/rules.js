// import Html from "slate-html-serializer";

import isUrl from "is-url"

export const BLOCK_TAGS = {
  p: "paragraph",
  blockquote: "quote",
  hr: "divider",
  h1: "heading",
  h2: "heading",
  h3: "heading",
  h4: "heading",
  a: "link",
  img: "image"
}

export const MARK_TAGS = {
  em: "italic",
  i: "italic",
  strong: "bold",
  b: "bold"
}

const squish = el => {
  el.innerHTML = el.innerText || el.textContent
  return el
}

export const RULES = [
  {
    deserialize(el, next) {
      const block = BLOCK_TAGS[el.tagName.toLowerCase()]
      if (!block) return
      switch (block) {
        case "paragraph": {
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes)
          }
        }
        case "quote": {
          return {
            object: "block",
            type: "quote",
            nodes: next(squish(el).childNodes)
          }
        }
        case "heading": {
          return {
            object: "block",
            type: "heading",
            nodes: next(squish(el).childNodes)
          }
        }
        case "image": {
          let imageSrc = el.getAttribute("src") || el.getAttribute("srcset")
          if (!isUrl(imageSrc)) return
          return {
            object: "block",
            type: "image",
            isVoid: true,
            data: { src: el.getAttribute("src") || el.getAttribute("srcset") }
          }
        }
        case "link": {
          return {
            object: "inline",
            type: "link",
            data: {
              href: el.getAttribute("href")
            },
            nodes: next(squish(el).childNodes)
          }
        }
        default:
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes)
          }
      }
    }
  },
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName]
      if (!mark) return
      return {
        object: "mark",
        type: mark,
        nodes: next(el.childNodes)
      }
    }
  }
]

//export const RULES = new Html({ rules });
