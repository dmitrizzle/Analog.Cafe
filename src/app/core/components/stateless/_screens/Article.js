// see /src/app/containers/Article
