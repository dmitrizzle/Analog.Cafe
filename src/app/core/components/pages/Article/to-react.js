import React from "react"
import { RULES } from "./rules" // What you normally feed slate-html-serializer

const rules = [
  ...RULES,
  // from slate-html-serializer
  {
    serialize(obj, children) {
      if (obj.object === "string") {
        return children.split("\n").reduce((array, text, i) => {
          if (i != 0) array.push(<br key={i} />)
          array.push(text)
          return array
        }, [])
      }
    }
  }
]

let serializeNode, serializeLeaf, serializeString, cruftNewline, addKey

const toReact = (value, options = {}) => {
  const { document } = value
  const elements = document.nodes.map(serializeNode).filter(Boolean)

  return elements
}

serializeNode = node => {
  if (node.object === "text") {
    const { leaves } = node
    return leaves.map(serializeLeaf)
  }

  const children = node.nodes.map(serializeNode)

  for (const rule of rules) {
    if (!rule.serialize) continue
    const ret = rule.serialize(node, children)
    if (ret === null) return
    if (ret) return addKey(ret)
  }

  throw new Error(`No serializer defined for node of type "${node.type}".`)
}

serializeLeaf = leaf => {
  const string = { object: "string", text: leaf.text }
  const text = serializeString(string)
  if (!leaf.marks) return leaf.text

  return leaf.marks.reduce((children, mark) => {
    for (const rule of rules) {
      if (!rule.serialize) continue
      const ret = rule.serialize(mark, children)
      if (ret === null) return
      if (ret) return addKey(ret)
    }

    throw new Error(`No serializer defined for mark of type "${mark.type}".`)
  }, text)
}

serializeString = string => {
  for (const rule of rules) {
    if (!rule.serialize) continue
    const ret = rule.serialize(string, string.text)
    if (ret) return ret
  }
}

cruftNewline = element =>
  !(element.nodeName === "#text" && element.nodeValue == "\n")

let key = 0

addKey = element => React.cloneElement(element, { key: key++ })

export default toReact
