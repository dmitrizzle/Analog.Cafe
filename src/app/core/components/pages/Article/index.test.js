import "jest-styled-components"

import { createMockStore } from "redux-test-utils"
import { shallowWithStore } from "enzyme-redux"
import React from "react"

import Article from "./"

it("Render Article, matches snapshot", () => {
  const store = createMockStore()
  const element = shallowWithStore(<Article />, store)
  expect(element).toMatchSnapshot()
})
