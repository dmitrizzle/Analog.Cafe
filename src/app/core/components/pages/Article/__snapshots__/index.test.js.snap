// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Article, matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Article) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "addSessionInfo": [Function],
      "article": undefined,
      "fetchArticlePage": [Function],
      "fetchListPage": [Function],
      "list": undefined,
      "setArticlePage": [Function],
      "user": undefined,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "addSessionInfo": [Function],
        "article": undefined,
        "fetchArticlePage": [Function],
        "fetchListPage": [Function],
        "list": undefined,
        "setArticlePage": [Function],
        "user": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "store": Object {
        "dispatch": [Function],
        "getAction": [Function],
        "getActions": [Function],
        "getState": [Function],
        "isActionDispatched": [Function],
        "isActionTypeDispatched": [Function],
        "subscribe": [Function],
      },
    },
  },
}
`;
